README.md
### Intro

@zooxsmart/lambda-middleware is a simple middleware implementation for AWS Lambda.

It uses async/await for handlers, no "callback hell".

This module also provides some basic middlewares.

### Example

```javascript
const middleware = require('@zooxsmart/lambda-middleware');
const CompanyMapper = require('./CompanyMapper');
const getParams = require('./getParams');

const handler = middleware.pipeline();

handler.use(getParams());
handler.use(middleware.halResponse().halEntity);

handler.handler(async event => {
  const mapper = new CompanyMapper();

  return mapper.fetch(event.pathParameters.id);
});

module.exports = { handler };
```

Here we have a get company example using some recommended lambda middlewares embedded in ```middleware.pipeline()```,
decorates the output with HAL links for the entity and a custom middleware ```getParam()``` to fetch database parameters.

### How it works

This module implements the middleware pipeline in a onion style.

The terminology "pipeline" is often used to describe the onion. One way of looking at the "onion" is as a queue, which is first-in-first-out (FIFO) in operation. This means that the first middleware on the queue is executed first, and this invokes the next, and so on (and hence the "next" terminology)

Each middleware receives the event and context, and pass to the handler only when wanting to hand off processing.

A middleware can return a response immediately, for example validate in input before handling it to the handler. In this case, the pipeline will not continue and the request will not even reach the handler.

### Pipeline

The pipeline will return the middleware engine.

Options:
```javascript
{
    useRecommended: true
}
```

* useRecommended: Will automatically add the following middlewares at the top:
    - doNotWaitEmptyEventLoop
    - errorHandler
    - eventParser

You can add other middlewares using the ```use(middleware)``` method and finally your handler with the ```handler(handler)``` method.

### Middlewares

#### Included (more on the way)

##### doNotWaitEmptyEventLoop(wait = false)
Will set ```context.callbackWaitsForEmptyEventLoop = wait;```

#### errorHandler
Will wrap the following middlewares and handler with a try/catch and return a formatted json response (recommended using http-errors). If not using http-errors, set ```expose``` to true so the message can be used. See http-errors for more details.

#### eventParser({ parseBody: true })
Will parse a body string into json and ensure the presence of pathParameters, queryStringParameters, requestContext and requestContext.authorizer.

#### halResponse({ path: null, collectionName: null, entityPath: null })
Will decorate an entity json with the _link.self from HAL and a collection with links, pagination, and it's entities.
* path: the path to include in the _link.self . Fallback to the current event.path if not provided
* collectionName: the collection name (duh) to be named in the _embedded
* entityPath: the entity path prefix. Will append the entity id

The middleware will use ```context.statusCode``` for the response with fallback to 200.

#### parameterStore(keyMap, destination, { ssmOptions: {}, expiryMs: 10 * 60 * 1000 })
Will fetch the keys from SSM Parameter Store, embed in the destination and cache the result for the specified expiryMs.
* keyMap: A map of key names and paths in SSM. Example:
{
    MYSQL_HOST: '/production/MYSQL_HOST',
    MYSQL_USER: '/production/MYSQL_USER',
    MYSQL_PASS: '/production/MYSQL_PASS',
}
* destination: the destination of the parameters. Example: ```process.env```, ```context```
* ssmOptions: Will be passed to the SSM constructor
* expiryMs: Will keep the parameters in the memory cache for expiryMs in milliseconds. Default is 10 minutes.


#### secretsManager(keyMap, destination, { secretsManagerOptions: {}, expiryMs: 10 * 60 * 1000 })
Will fetch the keys from AWS Secrets Manager, embed in the destination and cache the result for the specified expiryMs.
* keyMap: A map of key names and paths in Secrets Manager. Example:
{
    MYSQL_HOST: '/production/MYSQL_HOST',
    MYSQL_USER: '/production/MYSQL_USER',
    MYSQL_PASS: '/production/MYSQL_PASS',
}
* destination: the destination of the parameters. Example: ```context``` . context will be used if null
* secretsManagerOptions: Will be passed to the AWS.secretsmanager constructor
* expiryMs: Will keep the parameters in the memory cache for expiryMs in milliseconds. Default is 10 minutes.

#### Custom
You can create a custom middleware and even call another middleware, just follow the following signature and example:
```javascript
const middleware = require('../lambda-middleware');

module.exports = () => {
  return async (event, context, next) => {
    if (process.env.NODE_ENV !== 'development') {
      const params = {};
      await secretManager({ MYSQL: 'prod/mysql' })(event, params, (ev, ctx) => {
        process.env.MYSQL_HOST = ctx.MYSQL.host;
        process.env.MYSQL_USER = ctx.MYSQL.username;
        process.env.MYSQL_PASS = ctx.MYSQL.password;
      });
    }

    return next(event, context);
  };
};
```

```javascript
const middleware = require('../lambda-middleware');

module.exports = () => {
  return async (event, context, next) => {
    if (process.env.NODE_ENV !== 'development') {
      return middleware.parameterStore(
        {
          MYSQL_HOST: '/production/MYSQL_HOST',
          MYSQL_USER: '/production/MYSQL_USER',
          MYSQL_PASS: '/production/MYSQL_PASS',
        },
        process.env,
      )(event, context, next);
    }

    return next(event, context);
  };
};
```

***Warning*** Don't forget to call ```return next(event, context)``` at the end or the pipeline will not continue and will get no response.


CHANGELOG.md
## [1.1.1](https://github.com/zooxsmart/lambda-middleware/compare/1.1.0...1.1.1) (2019-09-12)


### Bug Fixes

* **halresponse:** use context.statusCode for response ([882c06e](https://github.com/zooxsmart/lambda-middleware/commit/882c06e))

# [1.1.0](https://github.com/zooxsmart/lambda-middleware/compare/1.0.1...1.1.0) (2019-09-09)


### Features

* **middleware:** add AWS Secrets Manager middleware ([dc54048](https://github.com/zooxsmart/lambda-middleware/commit/dc54048))

## [1.0.1](https://github.com/zooxsmart/lambda-middleware/compare/1.0.0...1.0.1) (2019-09-06)


### Bug Fixes

* fix deploy to npm ([1bd52b5](https://github.com/zooxsmart/lambda-middleware/commit/1bd52b5))

# 1.0.0 (2019-09-06)


### Features

* add initial files ([cee285a](https://github.com/zooxsmart/lambda-middleware/commit/cee285a))


src/pipeline.js
const doNotWaitEmptyEventLoop = require('./middlewares/doNotWaitEmptyEventLoop');
const errorHandler = require('./middlewares/errorHandler');
const eventParser = require('./middlewares/eventParser');

const app = (incomingOptions = {}) => {
  const options = {
    useRecommended: true,
    ...incomingOptions,
  };
  const middlewareList = [];
  let finalHandler;

  const run = ([first, ...others]) => {
    if (!first) {
      return finalHandler;
    }

    return async (event, context) => {
      return first(event, context, run(others));
    };
  };

  const instance = async (event, context) => {
    return run(middlewareList)(event, context);
  };

  instance.use = middleware => {
    middlewareList.push(middleware);
  };

  instance.handler = handler => {
    finalHandler = handler;
  };

  if (options.useRecommended) {
    instance.use(doNotWaitEmptyEventLoop());
    instance.use(errorHandler());
    instance.use(eventParser());
  }

  return instance;
};

module.exports = app;


src/middlewares/eventParser.js
/* eslint-disable no-param-reassign */
const createError = require('http-errors');

module.exports = incomingOptions => {
  const options = {
    parseBody: true,
    ...incomingOptions,
  };

  const httpParser = event => {
    event.pathParameters = event.pathParameters || {};
    event.queryStringParameters = event.queryStringParameters || {};
    event.requestContext = event.requestContext || {};
    event.requestContext.authorizer = event.requestContext.authorizer || {};
    if (options.parseBody) {
      try {
        event.body = typeof event.body === 'string' ? JSON.parse(event.body) : event.body;
      } catch (err) {
        throw new createError.BadRequest('Invalid JSON format');
      }
    }
  };

  return async (event, context, next) => {
    if (Object.prototype.hasOwnProperty.call(event, 'httpMethod')) {
      httpParser(event);
    }

    return next(event, context);
  };
};


src/middlewares/halResponse.js
const { response, hal } = require('@zooxsmart/lambda-util');

module.exports = incomingOptions => {
  const options = {
    path: null,
    collectionName: null,
    entityPath: null,
    ...incomingOptions,
  };

  const halEntity = async (event, context, next) => {
    const entity = await next(event, context);
    const resource = hal.toHal(entity, options.path || event.path);
    return response.createResponse(context.statusCode || 200, resource.toJSON());
  };

  const halCollection = async (event, context, next) => {
    const collection = await next(event, context);
    const resource = hal.toHalCollection(
      collection,
      options.collectionName,
      event.path,
      options.entityPath,
      event.queryStringParameters,
      context.collectionCount || null,
    );
    return response.createResponse(context.statusCode || 200, resource.toJSON());
  };

  return {
    halEntity,
    halCollection,
  };
};


src/middlewares/secretsManager.js
/* eslint-disable no-param-reassign,import/no-extraneous-dependencies */
const { promisify } = require('util');
const SecretsManager = require('aws-sdk/clients/secretsmanager');
const debug = require('debug')('lambda-middleware');

let secretsManager;
const cache = {
  expiration: new Date(0),
  items: {},
};

/**
 * @param {Object} keyMap
 * @param {Object} destination
 * @param {{
 *   secretsManagerOptions: Object,
 *   expiryMs: integer
 * }} incomingOptions
 * @returns {function(*=, *=, *): *}
 */
module.exports = (keyMap, destination, incomingOptions) => {
  const options = {
    secretsManagerOptions: {},
    expiryMs: 10 * 60 * 1000, // default expiry is 10 mins
    ...incomingOptions,
  };

  if (!keyMap) {
    throw new Error('You need to provide a keys map');
  }

  if (options.expiryMs <= 0) {
    throw new Error('You need to specify an expiry (ms) greater than 0, or leave it undefined');
  }

  const vars = Object.keys(keyMap);

  const createSecretsManagerClient = () => {
    if (secretsManager !== null) {
      secretsManager = new SecretsManager(options.secretsManagerOptions);
      // noinspection JSValidateTypes
      secretsManager.getSecretValue = promisify(secretsManager.getSecretValue);
    }
  };

  const reload = async () => {
    debug(`Loading cache keys: ${vars}`);

    createSecretsManagerClient();

    cache.items = {};
    await Promise.all(
      vars.map(async key => {
        const resp = await secretsManager.getSecretValue({ SecretId: keyMap[key] });
        const secret = JSON.parse(resp.SecretString || '{}');
        cache.items[key] = secret;
      }),
    );

    debug(`Successfully loaded cache keys: ${vars}`);
    const now = new Date();

    cache.expiration = new Date(now.getTime() + options.expiryMs);
  };

  return async (event, context, next) => {
    const now = new Date();
    if (now > cache.expiration) {
      await reload();
    }

    if (typeof destination !== 'object' || destination === null) {
      Object.assign(context, cache.items);
    } else {
      Object.assign(destination, cache.items);
    }

    return next(event, context);
  };
};


src/middlewares/parameterStore.js
/* eslint-disable no-param-reassign,import/no-extraneous-dependencies */
const { promisify } = require('util');
const SSM = require('aws-sdk/clients/ssm');
const debug = require('debug')('lambda-middleware');

let ssm;
const cache = {
  expiration: new Date(0),
  items: {},
};

/**
 * @param {Object} keyMap
 * @param {Object} destination
 * @param {{
 *   ssmOptions: Object,
 *   expiryMs: integer
 * }} incomingOptions
 * @returns {function(*=, *=, *): *}
 */
module.exports = (keyMap, destination, incomingOptions) => {
  const options = {
    ssmOptions: {},
    expiryMs: 10 * 60 * 1000, // default expiry is 10 mins
    ...incomingOptions,
  };

  if (!keyMap) {
    throw new Error('You need to provide a keys map');
  }

  if (typeof destination !== 'object' || destination === null) {
    throw new Error('Destination must be an object');
  }

  if (options.expiryMs <= 0) {
    throw new Error('You need to specify an expiry (ms) greater than 0, or leave it undefined');
  }

  const vars = Object.keys(keyMap);
  const keys = Object.values(keyMap);

  const createParameterStoreClient = () => {
    if (ssm !== null) {
      ssm = new SSM(options.ssmOptions);
      // noinspection JSValidateTypes
      ssm.getParameters = promisify(ssm.getParameters);
    }
  };

  const validate = (keys2, params) => {
    const missing = keys2.filter(k => params[k] === undefined);
    if (missing.length > 0) {
      throw new Error(`Missing keys: ${missing}`);
    }
  };

  const reload = async () => {
    debug(`Loading cache keys: ${vars}`);

    const req = {
      Names: keys,
      WithDecryption: true,
    };

    createParameterStoreClient();

    const resp = await ssm.getParameters(req);

    const params = {};
    resp.Parameters.forEach(p => {
      params[p.Name] = p.Value;
    });

    validate(keys, params);

    debug(`Successfully loaded cache keys: ${vars}`);
    const now = new Date();

    cache.expiration = new Date(now.getTime() + options.expiryMs);
    cache.items = params;
  };

  return async (event, context, next) => {
    const now = new Date();
    if (now > cache.expiration) {
      await reload();
    }

    vars.forEach(key => {
      destination[key] = cache.items[keyMap[key]];
    });

    return next(event, context);
  };
};
